### ISSUE TITLE
Triangulation issue with the WebGL polygon replay

### ISSUE THREAD

**Issue by jahow** [openlayers/openlayers#7818]
Hi,

I've encountered a few triangulation issues while experimenting with a Vector Tile renderer using WebGL, mostly with vast ocean features. I've tried my best to extract one problematic geometry and feed it as a single feature for easier debugging:
https://codepen.io/jahow/pen/NygPRE?editors=0010
![image](https://user-images.githubusercontent.com/10629150/36102717-705a7888-100d-11e8-8329-4825e94fdc8d.png)

This does not happen at every zoom level.

The issue will be more obvious on the POC WebGL renderer here: https://jahow.github.io/openlayers/examples/mapbox-vector-tiles.html
This example obviously uses an experimental fork from Openlayers.

@GaborFarkas I understand you've written the triangulation algorithm in the WebGL replays, do you think it should work with features from vector tiles or is that too far-fetched?

Thanks

------------------

- [x] I am submitting a bug or feature request, not a usage question. Go to https://stackoverflow.com/questions/tagged/openlayers for questions.
- [x] I have searched GitHub to see if a similar bug or feature request has already been reported.
- [x] I have verified that the issue is present in the latest version of OpenLayers (see 'LATEST' on https://openlayers.org/).
- [x] If reporting a bug, I have created a [CodePen](https://codepen.io) or prepared a stack trace (using the latest version and unminified code, so e.g. `ol-debug.js`, not `ol.js`) that shows the issue.

**Comment by GaborFarkas** [comment]
Hi Olivier,

Thank you for the bug report. I'm absolutely positive on the triangulation should work with vector tiles. This is a bug among many, which can-- and should be tracked down and eliminated. However, currently, I can only promise, I will dive into it when I'll get some free time on my hands. According to my current grants and projects, unfortunately, it will be no sooner than around July-August, 2018.

**Comment by jahow** [comment]
Ok, thanks. I'll try to better isolate the issue, that may give us a clue. For the record, could you summarize the algorithms used for triangulation and/or redirect me to documentation on this topic?

**Comment by GaborFarkas** [comment]
Of course. The algorithm is a modified [FIST](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.49.3013&rep=rep1&type=pdf). It first puts the ordered, valid, CCW oriented polygon's segments into a linked list (for practical reasons, it uses only pointers in segments, therefore every vertex is defined only once). Then, it triangulates the polygon with a penalizing technique:
![tri](https://user-images.githubusercontent.com/9150102/36373699-35f82832-1569-11e8-9dde-933559db8d7e.jpg)

The mathematical rules for simplicity and valid diagonals are coming from FIST. On the other hand, it was extended to handle self-intersections. Once it finds one, the algorithm inserts a Steiner point there, and continues with the triangulation.

At a first glance, based on the provided symptoms, the problem can easily be an unhandled case of edge collapsing. That is, OL's geometry simplifier simplifies the polygon on some zoom levels in such a way, that one or more vertices land on the polygon's edge or -- in worst case -- create a self-intersection in an unsuspected manner.

**Comment by ahocevar** [comment]
The WebGL polygon renderer has been removed.
